### ISSUE TITLE
[lua] unit_type.level seems broken

### ISSUE THREAD

**Issue by benoitkugler** [wesnoth/wesnoth#8456]
### Game and System Information

The Battle for Wesnoth version 1.17.25+dev (60ae0db6-Clean) x86_64
Running on Freedesktop SDK 23.08 (Flatpak runtime) x86_64
Distribution channel: Default
Linux

### Description of the bug

Using the following lua code

```lua
for id, ut in pairs(wesnoth.unit_types) do
     wesnoth.log("error", string.format("%s %d %d", id, ut.level, ut.__cfg.level))
end
```
shows that ut.level accessor is broken, whereas using the plain __cfg config works.

### Additional context

Here is a sample of what I got on the console 

20240225 16:21:06 error wml: Merman Brawler 0 1
20240225 16:21:06 error wml: Merman Citizen 0 0
20240225 16:21:06 error wml: Merman Entangler 0 3
20240225 16:21:06 error wml: Merman Fighter 0 1
20240225 16:21:06 error wml: Merman Hoplite 0 3
20240225 16:21:06 error wml: Merman Hunter 0 1
20240225 16:21:06 error wml: Merman Javelineer 0 3
20240225 16:21:06 error wml: Merman Netcaster 0 2
20240225 16:21:06 error wml: Merman Spearman 0 2
20240225 16:21:06 error wml: Merman Triton 0 3
20240225 16:21:06 error wml: Merman Warrior 0 2
20240225 16:21:06 error wml: Mudcrawler 0 0
20240225 16:21:06 error wml: Naga Dirkfang 0 1
20240225 16:21:06 error wml: Naga Fighter 0 1
20240225 16:21:06 error wml: Naga Guard 0 1
20240225 16:21:06 error wml: Naga High Guard 0 3
20240225 16:21:06 error wml: Naga Myrmidon 0 3
20240225 16:21:06 error wml: Naga Ophidian 0 2
20240225 16:21:06 error wml: Naga Ringcaster 0 2
20240225 16:21:06 error wml: Naga Shield Guard 0 2
20240225 16:21:06 error wml: Naga Sicarius 0 3
20240225 16:21:06 error wml: Naga Warrior 0 2
20240225 16:21:06 error wml: Naga Zephyr 0 3
20240225 16:21:06 error wml: Necromancer 0 3
20240225 16:21:06 error wml: Necrophage 0 2
20240225 16:21:06 error wml: Nibbler 0 0
20240225 16:21:06 error wml: Nightgaunt 0 3
20240225 16:21:06 error wml: Ogre 0 2
20240225 16:21:06 error wml: Orcish Archer 0 1
20240225 16:21:06 error wml: Orcish Assassin 0 1
20240225 16:21:06 error wml: Orcish Crossbowman 0 2
20240225 16:21:06 error wml: Orcish Grunt 0 1
20240225 16:21:06 error wml: Orcish Leader 0 1
20240225 16:21:06 error wml: Orcish Nightblade 0 3

**Comment by CelticMinstrel** [comment]
I looked into it, and it seem the issue is in iteration. If you do `wesnoth.unit_types[id].level` instead, it works.

**Comment by benoitkugler** [comment]
> I looked into it, and it seems the issue is in iteration. If you do `wesnoth.unit_types[id].level` instead, it works.

Alright, thank you. 
I somewhat need to iterate, but, for now, using `ut.__cfg.level` is a decent enough workaround for me.

**Comment by CelticMinstrel** [comment]
To clarify, you can do that _in the loop_ instead of using the loop variable `ut`, and it will still work.

In fact, if you just do a loop doing that at startup, normal iteration will work thereafter. The issue is that the loop catches a half-initialized unit type, whereas direct indexing makes sure the unit type is fully initialized.

So, putting this code anywhere in the game before your loop will make your loop read the correct value even if going off the `ut` loop variable:
```lua
for id, ut in pairs(wesnoth.unit_types) do
     wesnoth.unit_types[id].level
end
```

This is a bug in any case – iteration, like direct indexing, should make sure it's not returning half-initialized unit types.

**Comment by benoitkugler** [comment]
Ah indeed, I had missed your point, thank you for the clarification.

**Comment by CelticMinstrel** [comment]
I think that fix can be backported to 1.18 as well.
